### ISSUE TITLE
Information about deletions is printed all at once

### ISSUE THREAD

**Issue by razzolini** [WayneD/rsync#348]
## Description
Up to v3.2.3, information about deletions, such as `--info=del` (but also `--out-format=%o`, etc.), was printed in real time while rsync was deleting many files. Since v3.2.4, such information is instead printed in one or more "batches", whereas information about other operations (e.g. `--info=name`) is still printed in real time. I did quite a bit of experimentation to try and figure out the exact behavior, which seems to be the following:

1. at first, while rsync is only performing deletions, information about them is accumulated without printing it;
2. once rsync starts to perform non-deletion operations, it "flushes" the accumulated information about previous deletions by printing it all at once;
3. if rsync then goes back to performing multiple deletions in a row (for example, to delete files from another subdirectory in a recursive transfer), it stops printing and starts accumulating information again;
4. the cycle repeats from step 2, until the transfer ends;
5. at the end of the whole transfer, any remaining accumulated deletion information is printed all at once.

So, in the simplest case, when rsync is only performing deletions (e.g. because the source directory is empty), all information is printed at once at the end of the transfer.

## Reproduction
I've found that this issue is easier to observe if [`pv`](http://www.ivarch.com/programs/pv.shtml) is used to log the number of lines that rsync outputs over time, rather than looking at rsync's output directly.

Here is a shell script that should reproduce the issue, by creating 100000 files (you may want to adjust this number based on the speed of your machine) and then using rsync to delete them:

```bash
mkdir -p files/ empty/
for i in $(seq 100000); do
    > "files/$i.txt" # Create an empty file (faster than touch)
done
rsync --info=del --recursive --delete empty/ files/ |
    pv --numeric --interval 0.1 --line-mode --timer --bytes > /dev/null
```

On my machine, a typical output of this script with rsync 3.2.4 is

```
0.1803 0
0.2704 0
0.3605 0
0.4507 0
0.5408 0
0.6310 0
0.7211 0
0.8113 0
0.9000 59918
0.9818 100000
```

where each number in the first column shows the elapsed time, in seconds, and each corresponding number in the second column shows how many lines rsync had printed up to that time. So, we can see that rsync printed nothing for the better part of a second, and then printed all of the information about the 100000 deleted files at the end.

For comparison, a typical output of this script with rsync 3.2.3 is

```
0.1802 0
0.2055 2541
0.3090 15251
0.4071 27960
0.5005 40670
0.6053 55921
0.7640 68630
0.8054 73714
0.9081 91506
0.9847 100000
```

which shows that printing starts almost immediately.

Using the above script in combination with `git bisect`, I found out that the specific commit which introduced this issue is ead44adcd3eea08bfff2ba835943317706b76a73. I also confirmed that the issue still occurs with the latest commit from the master branch (0d8cc26044fe5c543f69206642666bfa7ee053f7).

## System info
My OS is Arch Linux, and `uname -a` prints

```
Linux home-desktop 5.18.16-arch1-1 #1 SMP PREEMPT_DYNAMIC Wed, 03 Aug 2022 11:25:04 +0000 x86_64 GNU/Linux
```

## Context
The reason I ran into this issue is that I use rsync to remove old backups from an external hard drive, because (at least in my case) it's faster than `rm -rf`. Deleting one of those backups can take a few minutes, so I like to see some indication of progress. However, rsync options such as `--info=progress2` don't show deletions. Instead, I use `--out-format=%o`, just so that rsync prints a line for each deletion ("del.", but the contents of the line don't matter), and I pipe this output into `pv` (much like in the above script) to display the elapsed time, the deletion rate, etc.

I understand that this is not a typical use case for rsync, and especially for its output, but I think that a way of seeing deletion progress in real time (be it through a simple line of output for each deletion, such as `--info=del`, or though some other progress indicator) might be generally useful.

**Comment by WayneD** [comment]
This is due to the buffered nature of the generator's checksum data & its messages sharing the same output stream. In an older rsync, the generator would stall waiting for the sender to read a bunch of checksums before it was able to work on deletions. Now, it can do more work with a larger message buffer when the socket is clogged up.

Feel free to use `--stderr=client` or `--stderr=all` to get the deletion messages output immediately (they will no longer be sent in the stream of data through the sender to be output by the receiver). They will be on stderr, though.

**Comment by razzolini** [comment]
Ah, sorry, I forgot to mention that I'd already tried `--stderr=all`, but it didn't help: deletion messages were still delayed in the same way. Now I've also tried `--stderr=client`, and this option doesn't help either (in fact, it doesn't even redirect deletion messages to stderr).

For completeness, other options that I tried and didn't help are `--delete-before`, `--delete-delay`, `--delete-after`, and all three `--outbuf` settings.
